File: citizensdk/android/native/src/main/kotlin/org/citizen/sdk/CitizenSdkModels.kt
package org.citizen.sdk

import java.math.BigInteger

enum class CitizenSdkLifecycle { CREATED, IMPORTING_STATE, STARTING, RUNNING, START_FAILED, STOPPED, DISPOSED }
enum class CitizenFinality { BEST, FINALIZED }
enum class CitizenWalletOrigin { CREATED, IMPORTED }
enum class CitizenWalletSignMode { HOT, COLD }
enum class CitizenSigningTransform { RAW, SUBSTRATE_SIGNING_PAYLOAD, BLAKE2_DOMAIN }
enum class CitizenExternalSignerTransport { QR_V1 }
enum class CitizenTransactionResolution { FINALIZED_SUCCESS, FINALIZED_FAILED, POOL_REJECTED }
enum class CitizenTransactionHistoryStatus { PENDING, IN_BLOCK, POOL_REJECTED, FINALIZED_SUCCESS, FINALIZED_FAILED }
enum class CitizenExecutionStatus { SUCCESS, FAILED, UNVERIFIED }

enum class CitizenCapabilityName {
    CHAIN_READ,
    TRANSACTION_BUILD,
    TRANSACTION_SUBMIT,
    TRANSACTION_VERIFY,
    WALLET_PROFILE,
    LOCAL_SIGNING,
    HARDWARE_VAULT,
    USER_AUTHENTICATION,
    HISTORY,
    BACKGROUND_SYNC,
}

enum class CitizenCapabilityReason {
    NONE,
    BUILD_UNSUPPORTED,
    DEVICE_UNAVAILABLE,
    HOST_DISABLED,
    ENGINE_NOT_RUNNING,
    DEPENDENCY_NOT_READY,
    USER_AUTHENTICATION_REQUIRED,
    VAULT_LOCKED,
    CHAIN_STARTING,
    CHAIN_UNSYNCED,
    STORAGE_UNAVAILABLE,
}

/** Exact unsigned 128-bit value with a Java-safe decimal representation. */
class CitizenU128(decimal: String) {
    val decimal: String
    @get:JvmSynthetic
    internal val low: Long
    @get:JvmSynthetic
    internal val high: Long

    init {
        // A canonical u128 has at most 39 decimal digits. Reject by O(1)
        // String length before regex or BigInteger performs proportional work.
        require(decimal.length in 1..MAX_DECIMAL_DIGITS) { "u128 is out of range" }
        require(decimal.matches(Regex("0|[1-9][0-9]*"))) { "u128 must be canonical unsigned decimal" }
        val value = BigInteger(decimal)
        require(value.bitLength() <= 128) { "u128 is out of range" }
        this.decimal = value.toString()
        low = value.and(MASK_64).toLong()
        high = value.shiftRight(64).toLong()
    }

    internal constructor(low: Long, high: Long) : this(
        unsignedLong(high).shiftLeft(64).or(unsignedLong(low)).toString(),
    )

    override fun equals(other: Any?): Boolean = other is CitizenU128 && decimal == other.decimal
    override fun hashCode(): Int = decimal.hashCode()
    override fun toString(): String = decimal

    companion object {
        private const val MAX_DECIMAL_DIGITS = 39
        private val MASK_64 = BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE)
        private fun unsignedLong(value: Long): BigInteger =
            BigInteger.valueOf(value and Long.MAX_VALUE).let {
                if (value < 0) it.setBit(63) else it
            }
    }
}

class CitizenBlockRef(
    hash: ByteArray,
    val number: String,
    val finality: CitizenFinality,
) {
    private val hashValue = hash.requireSize(32, "block hash")
    fun hash(): ByteArray = hashValue.clone()
}

class CitizenChainSyncStatus(
    val peerCount: String,
    val isSyncing: Boolean,
    val isUsable: Boolean,
    val best: CitizenBlockRef,
    val finalized: CitizenBlockRef,
)

class CitizenBlockHeader(
    val block: CitizenBlockRef,
    parentHash: ByteArray,
    stateRoot: ByteArray,
    extrinsicsRoot: ByteArray,
    digest: ByteArray,
) {
    private val parentValue = parentHash.requireSize(32, "parentHash")
    private val stateValue = stateRoot.requireSize(32, "stateRoot")
    private val extrinsicsValue = extrinsicsRoot.requireSize(32, "extrinsicsRoot")
    private val digestValue = digest.clone()
    fun parentHash(): ByteArray = parentValue.clone()
    fun stateRoot(): ByteArray = stateValue.clone()
    fun extrinsicsRoot(): ByteArray = extrinsicsValue.clone()
    fun digest(): ByteArray = digestValue.clone()
}

class CitizenBlockBody(
    val block: CitizenBlockRef,
    extrinsics: List<ByteArray>,
) {
    private val values = extrinsics.map(ByteArray::clone)
    fun extrinsics(): List<ByteArray> = values.map(ByteArray::clone)
}

class CitizenRuntimeContext(
    val block: CitizenBlockRef,
    val specVersion: Long,
    val transactionVersion: Long,
    metadata: ByteArray,
) {
    private val metadataValue = metadata.clone()
    fun metadata(): ByteArray = metadataValue.clone()
}

/** Explicit smoldot state transport only; this is not a legacy-app migration envelope. */
class CitizenChainState(
    val formatVersion: Long,
    val finalized: CitizenBlockRef,
    database: ByteArray,
) {
    init {
        require(formatVersion in 1..0xffff_ffffL) { "chain state format version must be nonzero uint32" }
        require(finalized.finality == CitizenFinality.FINALIZED) { "chain state anchor must be finalized" }
        require(database.isNotEmpty() && database.size <= 256 * 1024) { "chain state database is invalid" }
    }
    private val databaseValue = database.clone()
    fun database(): ByteArray = databaseValue.clone()
}

class CitizenCapabilityStatus(
    val name: CitizenCapabilityName,
    val reason: CitizenCapabilityReason,
    val supported: Boolean,
    val available: Boolean,
    val enabled: Boolean,
    val ready: Boolean,
)

class CitizenSdkCapabilities(
    val revision: String,
    statuses: List<CitizenCapabilityStatus>,
) {
    val statuses: List<CitizenCapabilityStatus> = statuses.toList()
}

class CitizenAccountBalance(
    val block: CitizenBlockRef,
    accountId: ByteArray,
    val freeFen: CitizenU128,
    val reservedFen: CitizenU128,
    val totalFen: CitizenU128,
) {
    private val accountIdValue = accountId.requireSize(32, "accountId")
    fun accountId(): ByteArray = accountIdValue.clone()
}

class CitizenAccountNonce(
    val bestBlock: CitizenBlockRef,
    accountId: ByteArray,
    val nonce: String,
) {
    private val accountIdValue = accountId.requireSize(32, "accountId")
    fun accountId(): ByteArray = accountIdValue.clone()
}

class CitizenFeeSnapshot(
    val bestBlock: CitizenBlockRef,
    val feeRateParts: Long,
    val minimumFeeFen: CitizenU128,
    val existentialDepositFen: CitizenU128,
)

class CitizenWalletAccount(
    val index: Long,
    accountId: ByteArray,
    val ss58Address: String,
    val name: String?,
    val createdAtMillis: String,
    val active: Boolean,
) {
    private val accountIdValue = accountId.requireSize(32, "accountId")
    fun accountId(): ByteArray = accountIdValue.clone()
}

class CitizenWalletProfile(
    val origin: CitizenWalletOrigin,
    val walletIndex: Long,
    val createdAtMillis: String,
    masterAccountId: ByteArray,
    activeAccountId: ByteArray,
    accounts: List<CitizenWalletAccount>,
) {
    private val masterValue = masterAccountId.requireSize(32, "masterAccountId")
    private val activeValue = activeAccountId.requireSize(32, "activeAccountId")
    val accounts: List<CitizenWalletAccount> = accounts.toList()
    fun masterAccountId(): ByteArray = masterValue.clone()
    fun activeAccountId(): ByteArray = activeValue.clone()
}

class CitizenWalletStateAccount(
    val signMode: CitizenWalletSignMode,
    val walletIndex: Long,
    val accountIndex: Long?,
    accountId: ByteArray,
    val ss58Address: String,
    val name: String,
    val createdAtMillis: String,
    val isDefault: Boolean,
) {
    private val accountIdValue = accountId.requireSize(32, "accountId")
    fun accountId(): ByteArray = accountIdValue.clone()
}

class CitizenWalletState(
    val revision: String,
    val hotProfile: CitizenWalletProfile?,
    accounts: List<CitizenWalletStateAccount>,
) {
    val accounts: List<CitizenWalletStateAccount> = accounts.toList()
    val defaultAccount: CitizenWalletStateAccount? get() = accounts.firstOrNull()
}

class CitizenSignature(signature: ByteArray) {
    private val value = signature.requireSize(64, "sr25519 signature")
    fun bytes(): ByteArray = value.clone()
}

class CitizenSigningIntent(
    accountId: ByteArray,
    payload: ByteArray,
    val transform: CitizenSigningTransform,
    domain: ByteArray = byteArrayOf(),
    val externalSignerTransport: CitizenExternalSignerTransport? = null,
    val opaqueAction: Int = 0,
    val ttlSeconds: Long = 120,
) {
    private val accountValue = accountId.requireSize(32, "accountId")
    private val payloadValue = payload.clone()
    private val domainValue = domain.clone()

    init {
        require(payloadValue.isNotEmpty() && payloadValue.size <= 16 * 1024 * 1024)
        require(opaqueAction in 0..0xffff && ttlSeconds in 1..300)
        require(
            (transform == CitizenSigningTransform.BLAKE2_DOMAIN && domainValue.size in 1..32) ||
                (transform != CitizenSigningTransform.BLAKE2_DOMAIN && domainValue.isEmpty()),
        )
    }

    fun accountId(): ByteArray = accountValue.clone()
    fun payload(): ByteArray = payloadValue.clone()
    fun domain(): ByteArray = domainValue.clone()
}

sealed class CitizenSigningOutcome(
    accountId: ByteArray,
    payloadHash: ByteArray,
) {
    private val accountValue = accountId.requireSize(32, "accountId")
    private val hashValue = payloadHash.requireSize(32, "payloadHash")
    fun accountId(): ByteArray = accountValue.clone()
    fun payloadHash(): ByteArray = hashValue.clone()

    class Completed(
        accountId: ByteArray,
        payloadHash: ByteArray,
        val signature: CitizenSignature,
    ) : CitizenSigningOutcome(accountId, payloadHash)

    class ExternalPending(
        accountId: ByteArray,
        payloadHash: ByteArray,
        val transport: CitizenExternalSignerTransport,
        val expiresAt: String,
        val sessionId: String,
        val transportRequest: String,
    ) : CitizenSigningOutcome(accountId, payloadHash)
}

sealed class CitizenDefaultAccountChangeOutcome(
    currentDefaultAccountId: ByteArray,
    payloadHash: ByteArray,
) {
    private val currentValue = currentDefaultAccountId.requireSize(32, "currentDefaultAccountId")
    private val hashValue = payloadHash.requireSize(32, "payloadHash")
    fun currentDefaultAccountId(): ByteArray = currentValue.clone()
    fun payloadHash(): ByteArray = hashValue.clone()

    class Completed(
        currentDefaultAccountId: ByteArray,
        payloadHash: ByteArray,
        val committedRevision: String,
    ) : CitizenDefaultAccountChangeOutcome(currentDefaultAccountId, payloadHash)

    class ExternalPending(
        currentDefaultAccountId: ByteArray,
        payloadHash: ByteArray,
        val transport: CitizenExternalSignerTransport,
        val expiresAt: String,
        val sessionId: String,
        val transportRequest: String,
    ) : CitizenDefaultAccountChangeOutcome(currentDefaultAccountId, payloadHash)
}

class CitizenExecution(
    val status: CitizenExecutionStatus,
    val reasonOrDispatchVariant: Long,
    val block: CitizenBlockRef?,
    val extrinsicIndex: Long?,
    val palletIndex: Long?,
    val errorIndex: Long?,
)

/** Safe projection of an SDK-owned opaque transaction preparation. */
class CitizenPreparedTransaction internal constructor(
    val preparationId: String,
    sourceAccountId: ByteArray,
    callDataHash: ByteArray,
    val bestBlock: CitizenBlockRef,
    val runtimeSpecNumber: Long,
    val transactionFormatNumber: Long,
    val nonce: String,
) {
    private val sourceValue = sourceAccountId.requireSize(32, "sourceAccountId")
    private val callHashValue = callDataHash.requireSize(32, "callDataHash")
    fun sourceAccountId(): ByteArray = sourceValue.clone()
    fun callDataHash(): ByteArray = callHashValue.clone()
}

sealed class CitizenTransactionExecution(
    val executionId: String,
    sourceAccountId: ByteArray,
    callDataHash: ByteArray,
) {
    private val sourceValue = sourceAccountId.requireSize(32, "sourceAccountId")
    private val callHashValue = callDataHash.requireSize(32, "callDataHash")
    fun sourceAccountId(): ByteArray = sourceValue.clone()
    fun callDataHash(): ByteArray = callHashValue.clone()

    class ExternalSigningPending internal constructor(
        executionId: String,
        sourceAccountId: ByteArray,
        callDataHash: ByteArray,
        val expiresAt: String,
        val qrRequest: String,
    ) : CitizenTransactionExecution(executionId, sourceAccountId, callDataHash)

    class Completed internal constructor(
        executionId: String,
        sourceAccountId: ByteArray,
        callDataHash: ByteArray,
        transactionHash: ByteArray,
        val resolution: CitizenTransactionResolution,
        val execution: CitizenExecution?,
        val poolRejectionReason: String?,
        replacementHash: ByteArray?,
    ) : CitizenTransactionExecution(executionId, sourceAccountId, callDataHash) {
        private val transactionHashValue = transactionHash.requireSize(32, "transactionHash")
        private val replacementHashValue = replacementHash?.requireSize(32, "replacementHash")
        fun transactionHash(): ByteArray = transactionHashValue.clone()
        fun replacementHash(): ByteArray? = replacementHashValue?.clone()
    }
}

/** Product-independent public projection of one SDK-submitted transaction. */
class CitizenTransactionHistoryRecord(
    executionId: ByteArray,
    sourceAccountId: ByteArray,
    callDataHash: ByteArray,
    transactionHash: ByteArray,
    val status: CitizenTransactionHistoryStatus,
    val block: CitizenBlockRef?,
    val execution: CitizenExecution?,
    replacementHash: ByteArray?,
    val createdAtMillis: String,
    val updatedAtMillis: String,
    val poolRejectionReason: String?,
) {
    private val executionIdValue = executionId.requireSize(16, "executionId")
    private val sourceValue = sourceAccountId.requireSize(32, "sourceAccountId")
    private val callHashValue = callDataHash.requireSize(32, "callDataHash")
    private val transactionHashValue = transactionHash.requireSize(32, "transactionHash")
    private val replacementHashValue = replacementHash?.requireSize(32, "replacementHash")
    fun executionId(): ByteArray = executionIdValue.clone()
    fun sourceAccountId(): ByteArray = sourceValue.clone()
    fun callDataHash(): ByteArray = callHashValue.clone()
    fun transactionHash(): ByteArray = transactionHashValue.clone()
    fun replacementHash(): ByteArray? = replacementHashValue?.clone()
}

/** Newest-first deterministic page from the SDK's execution-only store. */
class CitizenTransactionHistoryPage(
    val revision: String,
    records: List<CitizenTransactionHistoryRecord>,
    nextBeforeExecutionId: ByteArray?,
) {
    val records: List<CitizenTransactionHistoryRecord> = records.toList()
    private val nextValue = nextBeforeExecutionId?.requireSize(16, "nextBeforeExecutionId")
    fun nextBeforeExecutionId(): ByteArray? = nextValue?.clone()
}

internal fun ByteArray.requireSize(expected: Int, label: String): ByteArray {
    require(size == expected) { "$label must contain exactly $expected bytes" }
    return clone()
}

internal fun requireU64(value: String, label: String): ULong {
    require(value.length in 1..20 && value.matches(Regex("0|[1-9][0-9]*"))) {
        "$label must be canonical unsigned decimal"
    }
    return value.toULong()
}

/** 仅投影 Rust 模块位集合；依赖与非法组合由核心统一判定。 */
object CitizenSdkModules {
    const val WALLET = 1
    const val SIGNING = 2
    const val CHAIN = 4
    const val TRANSACTIONS = 8
    const val HISTORY = 16
    const val QR = 32
    const val FULL = 63
}

File: citizensdk/darwin/Sources/CitizenSDK/CitizenSDKModels.swift
import Foundation

/// 同一 Rust 核心的模块集合；这里只投影位值，组合合法性由核心统一校验。
public struct CitizenSDKModules: OptionSet, Sendable {
    public let rawValue: UInt32
    public init(rawValue: UInt32) { self.rawValue = rawValue }
    public static let wallet = Self(rawValue: 1)
    public static let signing = Self(rawValue: 2)
    public static let chain = Self(rawValue: 4)
    public static let transactions = Self(rawValue: 8)
    public static let history = Self(rawValue: 16)
    public static let qr = Self(rawValue: 32)
    public static let full: Self = [.wallet, .signing, .chain, .transactions, .history, .qr]

    internal var usesSecrets: Bool { !intersection([.wallet, .signing]).isEmpty }
}

public enum CitizenSDKLifecycle: UInt32, Sendable {
    case created = 1
    case importingState = 2
    case starting = 3
    case running = 4
    case startFailed = 5
    case stopped = 6
    case disposed = 7
}

public enum CitizenFinality: UInt32, Sendable { case best = 1, finalized = 2 }
public enum CitizenWalletOrigin: UInt32, Sendable { case created = 1, imported = 2 }
public enum CitizenWalletSignMode: UInt32, Sendable { case hot = 1, cold = 2 }
public enum CitizenSigningTransform: UInt32, Sendable {
    case raw = 1
    case substrateSigningPayload = 2
    case blake2Domain = 3
}
public enum CitizenExternalSignerTransport: UInt32, Sendable { case qrV1 = 1 }
@frozen public enum CitizenTransactionResolution: UInt32, Sendable {
    case finalizedSuccess = 1, finalizedFailed = 2, poolRejected = 3
}
public enum CitizenTransactionHistoryStatus: UInt32, Sendable {
    case pending = 1, inBlock = 2, poolRejected = 3, finalizedSuccess = 4, finalizedFailed = 5
}
public enum CitizenExecutionStatus: UInt32, Sendable { case success = 1, failed = 2, unverified = 3 }

public enum CitizenCapabilityName: UInt32, CaseIterable, Sendable {
    case chainRead = 1
    case transactionBuild = 2
    case transactionSubmit = 3
    case transactionVerify = 4
    case walletProfile = 5
    case localSigning = 6
    case hardwareVault = 7
    case userAuthentication = 8
    case history = 9
    case backgroundSync = 10
}

public enum CitizenCapabilityReason: UInt32, Sendable {
    case none = 0
    case buildUnsupported = 1
    case deviceUnavailable = 2
    case hostDisabled = 3
    case engineNotRunning = 4
    case dependencyNotReady = 5
    case userAuthenticationRequired = 6
    case vaultLocked = 7
    case chainStarting = 8
    case chainUnsynced = 9
    case storageUnavailable = 10
}

/// Exact unsigned 128-bit integer represented canonically in decimal.
public struct CitizenU128: Equatable, Hashable, Sendable, CustomStringConvertible {
    public let decimal: String
    internal let low: UInt64
    internal let high: UInt64

    public init(_ decimal: String) throws {
        try CitizenSDKChecks.require(!decimal.isEmpty && decimal.count <= 39, "u128 is out of range")
        try CitizenSDKChecks.require(decimal == "0" || (decimal.first != "0" && decimal.allSatisfy(\.isNumber)),
                                     "u128 must be canonical unsigned decimal")
        var high: UInt64 = 0
        var low: UInt64 = 0
        for scalar in decimal.unicodeScalars {
            guard scalar.value >= 48 && scalar.value <= 57 else {
                throw CitizenSDKError(.invalidArgument, "u128 must be canonical unsigned decimal")
            }
            let digit = UInt64(scalar.value - 48)
            let lowProduct = low.multipliedFullWidth(by: 10)
            let highProduct = high.multipliedReportingOverflow(by: 10)
            let highSum = highProduct.partialValue.addingReportingOverflow(lowProduct.high)
            let lowSum = lowProduct.low.addingReportingOverflow(digit)
            let highCarry = highSum.partialValue.addingReportingOverflow(lowSum.overflow ? 1 : 0)
            guard !highProduct.overflow && !highSum.overflow && !highCarry.overflow else {
                throw CitizenSDKError(.invalidArgument, "u128 is out of range")
            }
            high = highCarry.partialValue
            low = lowSum.partialValue
        }
        self.high = high
        self.low = low
        self.decimal = decimal
    }

    internal init(low: UInt64, high: UInt64) {
        self.low = low
        self.high = high
        self.decimal = CitizenU128.decimal(low: low, high: high)
    }

    public var description: String { decimal }

    private static func decimal(low: UInt64, high: UInt64) -> String {
        if high == 0 { return String(low) }
        var highPart = high
        var lowPart = low
        var digits: [UInt8] = []
        repeat {
            let highDivision = highPart.quotientAndRemainder(dividingBy: 10)
            let lowDivision = UInt64(10).dividingFullWidth((high: highDivision.remainder, low: lowPart))
            digits.append(UInt8(lowDivision.remainder) + 48)
            highPart = highDivision.quotient
            lowPart = lowDivision.quotient
        } while highPart != 0 || lowPart != 0
        return String(bytes: digits.reversed(), encoding: .ascii)!
    }
}

public struct CitizenBlockRef: Equatable, Sendable {
    public let hash: Data
    public let number: UInt64
    public let finality: CitizenFinality

    public init(hash: Data, number: UInt64, finality: CitizenFinality) throws {
        try CitizenSDKChecks.require(hash.count == 32, "block hash must contain exactly 32 bytes")
        self.hash = hash
        self.number = number
        self.finality = finality
    }
}

public struct CitizenChainSyncStatus: Equatable, Sendable {
    public let peerCount: UInt64
    public let isSyncing: Bool
    public let isUsable: Bool
    public let best: CitizenBlockRef
    public let finalized: CitizenBlockRef
}

public struct CitizenBlockHeader: Equatable, Sendable {
    public let block: CitizenBlockRef
    public let parentHash: Data
    public let stateRoot: Data
    public let extrinsicsRoot: Data
    public let digest: Data
}

public struct CitizenBlockBody: Equatable, Sendable {
    public let block: CitizenBlockRef
    public let extrinsics: [Data]
}

public struct CitizenRuntimeContext: Equatable, Sendable {
    public let block: CitizenBlockRef
    public let specVersion: UInt32
    public let transactionVersion: UInt32
    public let metadata: Data
}

/// Explicit smoldot state transport only; this is not a legacy-app migration envelope.
public struct CitizenChainState: Equatable, Sendable {
    public let formatVersion: UInt32
    public let finalized: CitizenBlockRef
    public let database: Data

    public init(formatVersion: UInt32, finalized: CitizenBlockRef, database: Data) throws {
        try CitizenSDKChecks.require(formatVersion > 0, "chain state format version must be nonzero")
        try CitizenSDKChecks.require(finalized.finality == .finalized,
                                     "chain state anchor must be finalized")
        try CitizenSDKChecks.require((1...256 * 1_024).contains(database.count),
                                     "chain state database is invalid")
        self.formatVersion = formatVersion
        self.finalized = finalized
        self.database = database
    }
}

public struct CitizenCapabilityStatus: Equatable, Sendable {
    public let name: CitizenCapabilityName
    public let reason: CitizenCapabilityReason
    public let supported: Bool
    public let available: Bool
    public let enabled: Bool
    public let ready: Bool
}

public struct CitizenSDKCapabilities: Equatable, Sendable {
    public let revision: UInt64
    public let statuses: [CitizenCapabilityStatus]
}

public struct CitizenAccountBalance: Equatable, Sendable {
    public let block: CitizenBlockRef
    public let accountID: Data
    public let freeFen: CitizenU128
    public let reservedFen: CitizenU128
    public let totalFen: CitizenU128
}

public struct CitizenAccountNonce: Equatable, Sendable {
    public let bestBlock: CitizenBlockRef
    public let accountID: Data
    public let nonce: UInt64
}

public struct CitizenFeeSnapshot: Equatable, Sendable {
    public let bestBlock: CitizenBlockRef
    public let feeRateParts: UInt32
    public let minimumFeeFen: CitizenU128
    public let existentialDepositFen: CitizenU128
}

public struct CitizenWalletAccount: Equatable, Sendable {
    public let index: UInt32
    public let accountID: Data
    public let ss58Address: String
    public let name: String?
    public let createdAtMillis: UInt64
    public let active: Bool
}

public struct CitizenWalletProfile: Equatable, Sendable {
    public let origin: CitizenWalletOrigin
    public let walletIndex: UInt32
    public let createdAtMillis: UInt64
    public let masterAccountID: Data
    public let activeAccountID: Data
    public let accounts: [CitizenWalletAccount]
}

/// One secret-free entry in the globally ordered hot/cold wallet catalog.
public struct CitizenWalletStateAccount: Equatable, Sendable {
    public let signMode: CitizenWalletSignMode
    public let walletIndex: UInt32
    public let accountIndex: UInt32?
    public let accountID: Data
    public let ss58Address: String
    public let name: String
    public let createdAtMillis: UInt64
    public let isDefault: Bool
}

/// Stable public wallet snapshot. The first account is the only default projection.
public struct CitizenWalletState: Equatable, Sendable {
    public let revision: UInt64
    public let hotProfile: CitizenWalletProfile?
    public let accounts: [CitizenWalletStateAccount]

    public var defaultAccount: CitizenWalletStateAccount? { accounts.first }
}

public struct CitizenSignature: Equatable, Sendable {
    public let bytes: Data
    internal init(_ bytes: Data) throws {
        try CitizenSDKChecks.require(bytes.count == 64, "sr25519 signature must contain exactly 64 bytes")
        self.bytes = bytes
    }
}

/// Product-independent signing input. Payload, domain and QR action are opaque
/// protocol bytes owned by the integrating application.
public struct CitizenSigningIntent: Equatable, Sendable {
    public let accountID: Data
    public let payload: Data
    public let transform: CitizenSigningTransform
    public let domain: Data
    public let externalSignerTransport: CitizenExternalSignerTransport?
    public let opaqueAction: UInt16
    public let ttlSeconds: UInt64

    public init(accountID: Data, payload: Data, transform: CitizenSigningTransform,
                domain: Data = Data(),
                externalSignerTransport: CitizenExternalSignerTransport? = nil,
                opaqueAction: UInt16 = 0, ttlSeconds: UInt64 = 120) throws {
        try CitizenSDKChecks.require(accountID.count == 32, "accountID must contain 32 bytes")
        try CitizenSDKChecks.require((1...16 * 1_024 * 1_024).contains(payload.count),
                                     "payload must contain 1...16 MiB bytes")
        try CitizenSDKChecks.require((1...300).contains(ttlSeconds), "ttlSeconds must be 1...300")
        try CitizenSDKChecks.require(
            (transform == .blake2Domain && (1...32).contains(domain.count)) ||
            (transform != .blake2Domain && domain.isEmpty),
            "signing transform/domain combination is invalid"
        )
        self.accountID = accountID
        self.payload = payload
        self.transform = transform
        self.domain = domain
        self.externalSignerTransport = externalSignerTransport
        self.opaqueAction = opaqueAction
        self.ttlSeconds = ttlSeconds
    }
}

@frozen public enum CitizenSigningOutcome: Equatable, Sendable {
    case completed(accountID: Data, payloadHash: Data, signature: CitizenSignature)
    case externalPending(accountID: Data, payloadHash: Data,
                         transport: CitizenExternalSignerTransport,
                         expiresAt: UInt64, sessionID: String, transportRequest: String)
}

@frozen public enum CitizenDefaultAccountChangeOutcome: Equatable, Sendable {
    case completed(currentDefaultAccountID: Data, payloadHash: Data, committedRevision: UInt64)
    case externalPending(currentDefaultAccountID: Data, payloadHash: Data,
                         transport: CitizenExternalSignerTransport,
                         expiresAt: UInt64, sessionID: String, transportRequest: String)
}

public struct CitizenExecution: Equatable, Sendable {
    public let status: CitizenExecutionStatus
    public let reasonOrDispatchVariant: UInt32
    public let block: CitizenBlockRef?
    public let extrinsicIndex: UInt32?
    public let palletIndex: UInt8?
    public let errorIndex: UInt8?
}

/// Secret-free summary for one non-persistent Core-owned transaction preparation.
public struct CitizenPreparedTransaction: Equatable, Sendable {
    public let preparationID: String
    public let sourceAccountID: Data
    public let callDataHash: Data
    public let bestBlock: CitizenBlockRef
    public let runtimeSpecNumber: UInt32
    public let transactionFormatNumber: UInt32
    public let nonce: UInt64
}

public struct CitizenTransactionExternalSigningPending: Equatable, Sendable {
    public let executionID: String
    public let sourceAccountID: Data
    public let callDataHash: Data
    public let expiresAt: UInt64
    public let qrRequest: String
}

public struct CitizenTransactionExecutionCompleted: Equatable, Sendable {
    public let executionID: String
    public let sourceAccountID: Data
    public let callDataHash: Data
    public let transactionHash: Data
    public let resolution: CitizenTransactionResolution
    public let execution: CitizenExecution?
    public let poolRejectionReason: String?
    public let replacementHash: Data?
}

@frozen public enum CitizenTransactionExecution: Equatable, Sendable {
    case externalSigningPending(CitizenTransactionExternalSigningPending)
    case completed(CitizenTransactionExecutionCompleted)
}

/// Product-independent public projection of one SDK-submitted transaction.
public struct CitizenTransactionHistoryRecord: Equatable, Sendable {
    public let executionID: String
    public let sourceAccountID: Data
    public let callDataHash: Data
    public let transactionHash: Data
    public let status: CitizenTransactionHistoryStatus
    public let block: CitizenBlockRef?
    public let execution: CitizenExecution?
    public let replacementHash: Data?
    public let createdAtMillis: UInt64
    public let updatedAtMillis: UInt64
    public let poolRejectionReason: String?
}

/// Newest-first deterministic page from the SDK's execution-only store.
public struct CitizenTransactionHistoryPage: Equatable, Sendable {
    public let revision: UInt64
    public let records: [CitizenTransactionHistoryRecord]
    public let nextBeforeExecutionID: String?
}
